dnssec-backend.go

```diff
@@ -0,0 +1,313 @@
+	ErrForgedRRsig          = errors.New("forged RRSIG header")
+	ErrRrsigValidationError = errors.New("RR doesn't validate against RRSIG")
+	ErrRrsigValidityPeriod  = errors.New("invalid RRSIG validity period")
+	ErrUnknownDsDigestType  = errors.New("unknown DS digest type")
+	ErrDsInvalid            = errors.New("DS RR does not match DNSKEY")
+	ErrDelegationChain      = errors.New("AuthChain has no Delegations")
+)
+
+type RRSet struct {
+	RrSet []dns.RR
+	RrSig *dns.RRSIG
+}
+
+type SignedZone struct {
+	Zone         string
+	Dnskey       *RRSet
+	Ds           *RRSet
+	ParentZone   *SignedZone
+	PubKeyLookup map[uint16]*dns.DNSKEY
+}
+
+type AuthenticationChain struct {
+	DelegationChain []SignedZone
+}
+
+func (r *RRSet) isSigned() bool {
+	return r.RrSig != nil
+}
+
+func (r *RRSet) isEmpty() bool {
+	return len(r.RrSet) < 1
+}
+
+func (r *RRSet) checkHeaderIntegrity(qname string) bool {
+	return r.RrSig != nil && r.RrSig.Header().Name != qname
+}
+
+func (z *SignedZone) checkHasDnskeys() bool {
+	return len(z.Dnskey.RrSet) > 0
+}
+
+func (z SignedZone) lookupPubKey(keyTag uint16) *dns.DNSKEY {
+	return z.PubKeyLookup[keyTag]
+}
+
+func (z SignedZone) addPubKey(k *dns.DNSKEY) {
+	z.PubKeyLookup[k.KeyTag()] = k
+}
+
+func removeRRSIGs(response *dns.Msg) *dns.Msg {
+	var filteredAnswers []dns.RR
+	filteredResponse := response.Copy()
+	for _, rr := range filteredResponse.Answer {
+		if _, isRRSIG := rr.(*dns.RRSIG); !isRRSIG {
+			filteredAnswers = append(filteredAnswers, rr)
+		}
+	}
+
+	filteredResponse.Answer = filteredAnswers
+	return filteredResponse
+}
+
+func setDNSSECdo(q *dns.Msg) *dns.Msg {
+	if q.IsEdns0() == nil {
+		q.SetEdns0(1024, true)
+	}
+	q.IsEdns0().SetDo()
+	return q
+}
+
+func newQuery(qname string, qtype uint16) *dns.Msg {
+	dnsMessage := &dns.Msg{
+		MsgHdr: dns.MsgHdr{
+			RecursionDesired: true,
+		},
+	}
+	dnsMessage.SetEdns0(4096, true)
+	dnsMessage.SetQuestion(qname, qtype)
+	return dnsMessage
+}
+
+func getRRset(qname string, qtype uint16, resolver Resolver, ci ClientInfo) (*RRSet, error) {
+	q := newQuery(qname, qtype)
+	r, err := doQuery(q, resolver, ci)
+	if err != nil || r == nil {
+		return nil, err
+	}
+	return extractRRset(r)
+}
+
+func extractRRset(r *dns.Msg) (*RRSet, error) {
+	result := &RRSet{RrSet: make([]dns.RR, 0)}
+	if r.Answer == nil {
+		return result, nil
+	}
+
+	result.RrSet = make([]dns.RR, 0, len(r.Answer))
+	for _, rr := range r.Answer {
+		switch t := rr.(type) {
+		case *dns.RRSIG:
+			result.RrSig = t
+		case *dns.DNSKEY, *dns.DS, *dns.A, *dns.AAAA:
+			result.RrSet = append(result.RrSet, rr)
+		}
+	}
+	return result, nil
+}
+
+func doQuery(q *dns.Msg, resolver Resolver, ci ClientInfo) (*dns.Msg, error) {
+	r, err := resolver.Resolve(q, ci)
+	if err != nil {
+		return nil, err
+	}
+	if r.Rcode == dns.RcodeNameError {
+		Log.Info("no such domain", "info", qName(r))
+		return nil, ErrNoResult
+	}
+	if r == nil || r.Rcode == dns.RcodeSuccess {
+		return r, err
```

Comment:
```suggestion
		return r, nil
```
`err` has already been checked and must be `nil` here.